ostd/src/mm/io.rs

```diff
@@ -199,6 +199,14 @@ pub struct KernelSpace;
 ///
 /// [valid]: core::ptr#safety
 unsafe fn memcpy(dst: *mut u8, src: *const u8, len: usize) {
+    // Note that the "volatile" keyword does not help data races between concurrent access. If data
+    // races happen, volatile loads are undefined behavior in Rust and target-specific behavior in
+    // LLVM. This should be fixed when we find a better way in the future (FIXME).
```

Comment:
Thanks for writing up such detailed comments, although I have reservations about the wording. It seems to suggest that we are doing things in a wrong or dangerous way. If I understand it correctly, your main concern is that we may use this `memcpy` to write memory that is being read by other threads, or read memory that is being written by other, leading to data races. Data races are undefined behavior (UB) in Rust. So the code causes UB.

But I don't think so because I think Asterinas should make an amendament to Rust's UB definition. I call it _UB with untyped memory exception_. Data races is on the top of [the official-yet-quite-informal UB list of Rust](https://doc.rust-lang.org/reference/behavior-considered-undefined.html). But data races are harmless to untyped memory as one can only read/write `Pod` values from/into it.

Back on this `memcpy` function, allowing data races is not a sin when the buffer refers to untyped memory. So whether the "volatile" keyword help data races or not is less relevant. What should be done is to amend the safety section of the function Rust doc.

```rust
/// Copies `len` bytes from `src` to `dst`.
///
/// # Safety
///
/// - The source memory range of `(src .. src + len)` must refer to either typed or untyped memory, not mixed.
/// So is the destination memory range of `(dst ... dst + len)`.
/// - If the source or destination memory range refers to typed memory, then it must be [valid].
/// - If both the source and destination memory ranges are typed memory, then they must not overlap.
/// - If the source or destination memory range refers to untyped memory, then it must remain untyped memory during the entire call to `memcpy`.
///
/// [valid]: core::ptr#safety
unsafe fn memcpy(dst: *mut u8, src: *const u8, len: usize) {
```

Reply:
> Thanks for writing up such detailed comments, although I have reservations about the wording. It seems to suggest that we are doing things in a wrong or dangerous way. If I understand it correctly, your main concern is that we may use this `memcpy` to write memory that is being read by other threads, or read memory that is being written by other, leading to data races. Data races are undefined behavior (UB) in Rust. So the code causes UB.

My point is that it causes UB, so the fact should at least be explicitly documented. If you like, the wording can be revised to state the fact in a more positive way, since we're just accepting the UB.

> But I don't think so because I think Asterinas should make an amendament to Rust's UB definition. I call it _UB with untyped memory exception_. Data races is on the top of [the official-yet-quite-informal UB list of Rust](https://doc.rust-lang.org/reference/behavior-considered-undefined.html). But data races are harmless to untyped memory as one can only read/write `Pod` values from/into it.

To me, this is not a valid argument. In this paragraph, you state your conclusion that we can safely exclude data races on untyped memory from UB, but you don't provide any evidence for your conclusion.

You say we can do it safely, but _why_ can we do it safely if the official Rust documentation says we should not? Conversely, when I told you that doing this is UB and we should avoid this UB, I provided you with [_lots of_ evidence](https://github.com/asterinas/asterinas/pull/895#discussion_r1656342489) (e.g., a number of references to documentation and discussions in the Rust and LLVM communities, and even a concrete example that leads to mis-optimizations).

Please _don't_ say that the Rust documentation only talks about typed memory, but we're dealing with untyped memory. _If_ that is the case, there won't be as much [open discussion](https://github.com/rust-lang/unsafe-code-guidelines/issues/152) in the Rust community as I mentioned in my previous comments.

_The real problem_ is that, by the definition of UB, a UB means that the entire behavior of the program is undefined, _not_ just the statement that triggers the UB gives unexpected results. So UBs cannot be distinguished by whether they operate on typed or untyped memory, which simply makes no sense. Why not? Because the UB is there for compiler optimizations. The compiler assumes that the UB doesn't exist in order to optimize the program, and the behavior of the whole program cannot be predicted if it turns out that the compiler just made a wrong assumption.

In addition to [the example](https://github.com/asterinas/asterinas/pull/895#discussion_r1657982603) I shared with you earlier, you can also look at a number of examples in this paper [How to miscompile programs with "benign" data races](https://www.usenix.org/legacy/events/hotpar11/tech/final_files/Boehm.pdf). I quote the main conclusion of the paper below:

>> Nonetheless many programmers clearly believe, along with [15] that certain kinds of data races can be safely ignored in practice because they will produce expected results with all reasonable implementations. Here we show that all kinds of C or C++ source-level “benign” races discussed in the literature can in fact lead to incorrect execution as a result of perfectly reasonable compiler transformations, or when the program is moved to a different hardware platform.

So, since the "volatile" keyword prohibits compiler optimizations, should the problem be solved? Unfortunately, the answer is still somewhat mixed.
 - Yes, in practice. The Linux kernel uses the "volatile" keyword to implement the [`READ_ONCE` macro](https://elixir.bootlin.com/linux/v6.9.8/source/include/asm-generic/rwonce.h#L44). This macro can be used to access a variable that might be concurrently modified by another thread (using something like `WRITE_ONCE`). Thus, if there are compilers that produce incorrect results for volatile but racy memory accesses, such compilers cannot be used to compile the Linux kernel.
 - No, in theory. The C/C++ standards and the Rust documentation still consider all data races to be UBs, regardless of whether the "volatile" keyword is used. So as far as the standards are concerned, the compiler is allowed to generate programs that do arbitrarily anything after the UB occurs. As I mentioned in the previous comments, even the LLVM community refuses to give a stronger guarantee.

In the future, this problem will be fully solved if one of the following happens:
 - The C/C++ standards exclude concurrent volatile memory accesses from the definition of data races. There is a proposal [P1382R1](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1382r1.pdf) tries to do this in 2019, but unfortunately it hasn't been accepted after 5 years. If some similar proposal gets accepted in the future, the LLVM/Rust community should immediately follow the C/C++ standards.
 - The LLVM community agrees to guarantee that volatile memory loads never return `undef`, even if data races occur. The Rust community may follow suit to make the same guarantee. However, in this case, the Rust community may not follow either, since LLVM is not the only Rust backend.
 - [New APIs](https://github.com/rust-lang/unsafe-code-guidelines/issues/321) designed specifically for this are added to the Rust core library. I also mentioned this solution in my previous comments, but it will not happen soon either, as the new APIs are currently in the pre-pre-RFC stage for two years.

Reply:
> It seems to suggest that we are doing things in a wrong or dangerous way.

I revised the comments to focus more on explaining why using the keyword "volatile" here is not that dangerous, even if it is UB. Does this address your concerns?

Reply:
Thanks for all the references you provide. Very useful. I now read most of them.

Yes, in the most strict terms, even using `volatile`-version of pointer read, write, and copy memory API to access untyped memory is considered UB. But what is the real harm of accessing untyped memory concurrently with `volatile_copy_memory `? And what is the better alternative to volatile access? Writing assembly code? I believe our implementation is good enough and correct enough in practice.

> If data races happen, volatile loads are undefined behavior ... This should be fixed ...

You must agree that accessing untyped memory concurrently via volatile API is not the same kind of UB as, let's say, dereferencing dangling pointers, right? In fact, the [open discussion](https://github.com/rust-lang/unsafe-code-guidelines/issues/152) that you mentioned even proposed to remove our use case from UB.

> a) the documentation be changed not to designate concurrent external modification of memory locations that a Rust program only accesses as volatile to be UB and b) to state in the Unsafe Code Guidelines that it's legitimate to use volatile accesses to access memory that a thread of execution external to the Rust program might change concurrently.

This is why I have strong opinions again the comments. The comments are written in such a tone that sounds like that we have a BIG soundness hole here; in fact, it is not.

In addition, there are multiple locations that invoke `ptr::volatile_read` or `ptr::volatile_write` for untyped memory. By the standard as you propose, all these locations should be commented with FIXME. I think seeing these comments would be a distraction to readers rather than an assistance.

Reply:
The Asterinas Book has a placeholder page called Soundness Analysis. I have long planned to write up the soundness story of `ostd`, but haven't got the time to do so. I think our discussion in this PR and the references you provide would be very useful materials to it.

If you can't think of a better location for your comments, then we can just leave them in the current location and remove them in the future when the Soundness Analysis page has done its job properly.

Reply:
Thanks for your detailed comments! The points in your comments make sense to me and I also agree with your statement that my previous tone was a bit confusing.

As for the location, I actually think that putting the story in Asterina's book is pretty good. As you say, I write something like "for more details, see some random GitHub discussion" in the code comments to avoid making the comments unnecessarily large and interrupting the code reader. But this is not perfect, we'd better write the soundness story formally in some formal place. Asterina's book is a wonderful choice.

> If you can't think of a better location for your comments, then we can just leave them in the current location and remove them in the future when the Soundness Analysis page has done its job properly.

So I vote this.

Reply:
Great. Then I will merge this PR.